lib/Conversion/LLHDToLLVM/LLHDToLLVM.cpp

```diff
@@ -0,0 +1,1427 @@
+    // malloc function signature: (i64 %size) -> i8* %pointer
+    auto mallocSigFuncTy =
+        LLVM::LLVMType::getFunctionTy(i8PtrTy, {i64Ty}, false);
+    auto mallFunc =
+        getOrInsertFunction(module, rewriter, "malloc", mallocSigFuncTy);
+
+    //! get or insert library call definition
+    // alloc_signal function signature: (i8* %state, i8* %sig_name, i8*
+    // %sig_owner, i32 %value) -> i32 %sig_index
+    auto allocSigFuncTy = LLVM::LLVMType::getFunctionTy(
+        i32Ty, {i8PtrTy, i32Ty, i8PtrTy, i8PtrTy, i64Ty}, false);
+    auto sigFunc =
+        getOrInsertFunction(module, rewriter, allocCall, allocSigFuncTy);
+
+    // get or insert alloc_proc library call definition
+    auto allocProcFuncTy = LLVM::LLVMType::getFunctionTy(
+        voidTy, {i8PtrTy, i8PtrTy, i8PtrTy}, false);
+    auto allocProcFunc =
+        getOrInsertFunction(module, rewriter, "alloc_proc", allocProcFuncTy);
+
+    Value initStatePtr = initFunc.getArgument(0);
+
+    // get builder with insertion point before the init function terminator
+    OpBuilder initBuilder =
+        OpBuilder::atBlockTerminator(&initFunc.getBody().getBlocks().front());
+
+    // use the instance name to retrieve the instance from the state
+    auto ownerName = instOp.name();
+    //! get or create owner name string
+    Value owner;
+
+    auto parentSym =
+        module.lookupSymbol<LLVM::GlobalOp>("instance." + ownerName.str());
+    if (!parentSym) {
+      owner = LLVM::createGlobalString(
+          rewriter.getUnknownLoc(), initBuilder, "instance." + ownerName.str(),
+          ownerName.str() + '\0', LLVM::Linkage::Internal,
+          typeConverter.getDialect());
+      parentSym =
+          module.lookupSymbol<LLVM::GlobalOp>("instance." + ownerName.str());
+    } else {
+      owner = getGlobalString(rewriter.getUnknownLoc(), initBuilder,
+                              typeConverter, parentSym);
+    }
+
+    if (auto child = module.lookupSymbol<EntityOp>(instOp.callee())) {
+      // walk over the unit and generate mallocs for each one of its signals
+      // index of the signal in the unit's signal table
+      int initCounter = 0;
+      child.walk([&](Operation *op) -> WalkResult {
+        if (auto sigOp = dyn_cast<SigOp>(op)) {
+          // get index constant of the signal in the unit's signal table
+          auto indexConst = initBuilder.create<LLVM::ConstantOp>(
+              rewriter.getUnknownLoc(), i32Ty,
+              rewriter.getI32IntegerAttr(initCounter));
+          initCounter++;
+
+          //! add signal allocation to the init function
+          // clone and insert init's defining operation (assmued to be a
+          // constant op)
+          auto initDef =
+              initBuilder.insert(sigOp.init().getDefiningOp()->clone())
+                  ->getResult(0);
+          // malloc required space
+          int size =
+              std::ceil(sigOp.init().getType().getIntOrFloatBitWidth() / 8.0);
```

Comment:
Instead of floating point, please use functions like `divideCeil` in llvm's MathExtras.h